Fully deprecate AutoGPTQ and AutoAWQ for GPT-QModel

Qubitium opened the pull request:
Remove autogptq clutter and autogptq related configs that are not worth adding backward compat. 

GPTQModel has a slight project name change (pypi package and import name stays the same) to GPT-QModel with `-` as we now have added `awq`/AutoAWQ into our repo and will be making pr soon to address awq loading using GPT-QModel. 

`GPTQConfig` has the most important changes in this PR:

```python

# New GPTQConfig Property. Applicable for sister Peft/Optimum PRs
act_group_aware (`bool`, *optional*, defaults to `True`):
    Use GAR (group aware activation order) during quantization. Has measurable positive impact on quantization
    quality. Only applicable when `desc_act = False`. Will forced to be `False` when `desc_act = True`.
    
    
# Removed GPTQConfig Properties:
use_cuda_fp16
use_exllama
exllama_config
```

The 3 removed properties are all related `kernel` selection. These 3 are a hot potatoe mess and legacy from autogptq. GPT-QModel uses unified `backend` (existing) property to select kernels.  There were compat codes written to `convert`  these 3 properties to `backend` behind the scenes in 2024 but no longer relevant for 2025. 

Note: 

* Transformers/Optimum/Peft  CI tests should never check for `kernel.QUANT_TYPE` (str). GPTQ-QModel will return best performing kernel for the relevant module and it may be different per module due to in/out features and other gptq/module properties in relation to device type + dtype + many factors. 
* CI tests should only assert check for `kernel.QUANT_TYPE` if the test specifies a specific kernel via `backend` selection.

Qubitium committed:
fully deprecate autogptq

Qubitium committed:
remove use_cuda and use_exllama toggles are fully deprecated in gptqmodel

Qubitium committed:
format

Qubitium committed:
add `act_group_aware` property

Qubitium committed:
Merge branch 'main' into gptqmodel

ZX-ModelCloud committed:
fix QUANT_TYPE assert

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
Merge remote-tracking branch 'origin/gptqmodel' into gptqmodel

Rocketknight1 commented:
cc @mekkcyber for quantization

Qubitium committed:
Merge branch 'main' into gptqmodel

Qubitium committed:
Merge branch 'main' into gptqmodel

Qubitium committed:
format

Qubitium committed:
Merge branch 'main' into gptqmodel

Qubitium committed:
Merge branch 'main' into gptqmodel

Qubitium committed:
Merge branch 'main' into gptqmodel

LRL2-ModelCloud committed:
mod awq import

LRL2-ModelCloud committed:
remove autoawq fuse support

LRL2-ModelCloud committed:
remove remove autoawq.config fuse

LRL2-ModelCloud committed:
cleanup

LRL2-ModelCloud committed:
remove awq fuse test

LRL2-ModelCloud committed:
fix import

LRL2-ModelCloud committed:
use gptqmodel

LRL2-ModelCloud committed:
cleanup

LRL2-ModelCloud committed:
remove get_modules_to_fuse

LRL2-ModelCloud committed:
mod require_auto_awq -> require_gptqmodel

MekkCyber commented:
Hi @Qubitium ! Thanks a lot for working on this! Quick question, what do you mean by AutoAWQ being part of GPT-QModel now? Did you integrate the entire library (including the transformers dependency, like AutoAWQ does), or did you just port over the linear layers, kernels, and related components?

LRL2-ModelCloud committed:
convert vertion to checkpoint_format

LRL2-ModelCloud committed:
check is_gptqmodel_available

LRL2-ModelCloud committed:
revert modules_to_not_convert

Qubitium commented:
> Hi @Qubitium ! Thanks a lot for working on this! Quick question, what do you mean by AutoAWQ being part of GPT-QModel now? Did you integrate the entire library (including the transformers dependency, like AutoAWQ does), or did you just port over the linear layers, kernels, and related components?

Long story short. We folded AutoAWQ into GPT-QModel in multiple stage over the past few weeks. Stage 1. Simple/Directly port/copy  the AutoAWQ code over.  Stage 2. Refractor. Stage 3. Fixed bugs, add new kernels. Major refractor to align with new internal life cycle in GPT-QModel v5.0+. So we are current post Stage 3 where GPT-QModel base retains minimal original AutoAWQ code. Most AutoAWQ code has been refractored away. 

Major Changes vs AutoAWQ:

1. New kernels. We have 2 new kernels for AWQ (`AwqTorch` pure torch based, and `AwqTorchFused` which is cpu optimized based on work by Intel @jiqing-feng. 
2. Plan to add 3rd new AWQ kernel based on Bitblas as most gptq kernels are compatbile with AWQ with some small changes. Marlin kernel also sycned with gptq Marlin kernel for updated Marlin fixes/otpimizations via vllm port. 
3. QuantLinear code have been rewritten/refractored. 
4. Quant logic is new due to GPT-QModel 5.0+ life cycle which is not compatible with AutoAWQ. 

HF eco system compat:

Work on `Peft` integration is happening in a parallel pr by @LRL2-ModelCloud https://github.com/huggingface/peft/pull/2917 with coordination from @BenjaminBossan  https://github.com/huggingface/peft/issues/2342#issuecomment-3547697772

The `Peft` pr will need to co-exist concurrently with this PR due to interdependency. 

We will hold off `Optimum` change last if we  can help it, or may have to parallel a 3rd Pr to `Optimum` as well if inter-dependency causes trouble there as well. 

Final goal of the 2 Prs is to remove dead AutoGPTQ code (no one uses it or should use frankly) and almost dead AutoAWQ (repo in readonly and no longer accepting bug fixes or new model support).  Compat of model loading of old models that usd these two packages will be maintained.

LRL2-ModelCloud committed:
pass bits, sym, desc_act

Qubitium commented:
> About GPTQ-Model, will it be possible to create awq quants for newer models that are compatible with other frameworks (e.g vllm) just like autoawq did ?

Our quantized models are more compatible with vllm/SGLang than ones quantized with Optimum or AutoAWQ

SGLang/vLLM compat is a number 1 target/design from day one so 100% yes.

LRL2-ModelCloud committed:
fix awqconfig init

LRL2-ModelCloud committed:
fix wrong args

jiqing-feng commented:
It's a good chance to deprecate autoawq as it's archived. I suppose the best way is to go upstream to the transformer's main codes, just like we did in the Autogptq replacement. For example, the [IPEX linear in AutoAWQ](https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/awq.py#L212) is out-of-date, we need a new implementation for it. The new linear implementation is TorchFusedLinear in gptqmodel.

LRL2-ModelCloud committed:
fix ipex

LRL2-ModelCloud committed:
mod ipex version check

LRL2-ModelCloud committed:
cleanup

LRL2-ModelCloud committed:
fix awq_linear

LRL2-ModelCloud committed:
remove self.exllama_config = exllama_config

LRL2-ModelCloud committed:
cleanuo

LRL2-ModelCloud committed:
Revert "cleanuo"

This reverts commit 90019c6fc4f7a617ed9db482a42ecd1cd07f9108.

LRL2-ModelCloud committed:
update is_trainable

LRL2-ModelCloud committed:
cleanup

SunMarc commented on src/transformers/integrations/awq.py:
```diff
@@ -124,35 +124,35 @@ def replace_with_awq_linear(
+            from gptqmodel.nn_modules.qlinear.awq_gemm import AwqGEMMQuantLinear
 
-            target_cls = WQLinear_GEMM
+            target_cls = AwqGEMMQuantLinear
```
As I said, we might have replace this path by one handled by `kernels` at some point

Qubitium replied:
That's fine. At that point in the future, HF staff can override the auto kernel selection code from gpt-qmodel and return specific `AwqGEMM` from `kernels`. It will be a clean override and requires no changes from gpt-qmodel. My point is that it is unreasonable to burden our task further by imposing a future optional requirement that that does not resolve the issue now.

SunMarc reviewed (commented):
Thanks, left a couple of comments. As I said, I'm happy to see that you are willing to fill the hole left by AutoAWQ and eager to see this PR merged. However, note that maybe in the future, we will add a default working path for GEMM if gptq-model is not installed. As those libraries depends on kernels that requires to deal with building + distribution for each new version of torch, we never know when this will suddenly stop. 
Left a couple of comments. Also maybe it will be better to split this PR into 2: one for gptq and one for awq ?

Qubitium commented on src/transformers/quantizers/quantizer_awq.py:
```diff
@@ -129,12 +129,6 @@ def _process_model_before_weight_loading(
-
-            model = fuse_awq_modules(model, self.quantization_config)
-            model._awq_is_fused = True  # TODO: consider storing this flag in model.config instead
-
```
Fusing has no place in AWQ. QKV, MLP fusing should be done at an upper level just like what vLLM and SGLang is doing. This is wrong code/logic for 2025. AutoAWQ did it to squeeze some inference perpformance but if you look at the code, it is a hot mess of static model class mapping and unmtainable. The number of new models coming out and emergence of SGLang, vLLM make this obsolete. It is my understanding that v5 HF is going to be used for SGLang/vLLM loading model foudnation so that makes it even more so. 

For those users that depend on `AutoAWQ` fusing, they need to choose. We are not gonna spend vaulable energy supporting dead code.

Qubitium commented on src/transformers/integrations/awq.py:
```diff
@@ -124,35 +124,35 @@ def replace_with_awq_linear(
+    if not is_gptqmodel_available():
         raise ValueError(
-            "AWQ (either `autoawq` or `llmawq`) is not available. Please install it with `pip install autoawq` or check out the installation guide in https://github.com/mit-han-lab/llm-awq"
+            "AWQ (either `llmawq`) is not available. Please install it with `pip install gptqmodel` or check out the installation guide in https://github.com/mit-han-lab/llm-awq"
```
I will try to make this as clean as possible to reach that v5 goal. We are removig 75% of the code, not adding. Other than `fusing`, the features are not deprecated and only improved with zero compt issues. New kernels, bettter hw compat, faster kernels, and even bug fixes. Current awq kernels are `fp16` only and failed all our `bf16` kernel output quality tests. We will make sure users do not execute in bf16 or at least warn when this happens (loading model and executing in bf16 when awq kernels are validated for fp16 only).

Qubitium commented on src/transformers/quantizers/quantizer_gptq.py:
```diff
@@ -22,7 +22,7 @@
     from ..modeling_utils import PreTrainedModel
 
-from ..utils import is_auto_gptq_available, is_gptqmodel_available, is_optimum_available, is_torch_available, logging
+from ..utils import is_gptqmodel_available, is_optimum_available, is_torch_available, logging
```
https://github.com/huggingface/peft/pull/2917#discussion_r2551674217

GPTQ changes are minimal and mostly cosmetic. But this PR is required for https://github.com/huggingface/peft/pull/2917#discussion_r2551674217 due to interdependency.

Qubitium committed:
merge

Qubitium commented:
@SunMarc @MekkCyber  Hold off on review. I will ping once ready. I need to remove more code related to fuse and kernel selection.

Qubitium committed:
remove fused

ZX-ModelCloud committed:
call hf_select_quant_linear_v2()

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
Remove the "version" field from AwqConfig

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
Add torch_fused inferencefix test_gptq test

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
fix test_awq

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
fix test_awq

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
fix AwqConfig

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

Qubitium committed:
Merge branch 'main' into gptqmodel

ZX-ModelCloud committed:
call hf_select_quant_linear_v2()

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
Merge remote-tracking branch 'origin/gptqmodel' into gptqmodel

ZX-ModelCloud committed:
remove auto_awq

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
fix typo

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

Qubitium committed:
Merge branch 'main' into gptqmodel

ZX-ModelCloud committed:
Compatible with legacy field: checkpoint_format

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
Compatible with legacy field: checkpoint_format

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
format

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

SunMarc commented on src/transformers/quantizers/quantizer_awq.py:
```diff
@@ -129,36 +87,23 @@ def _process_model_before_weight_loading(
+        # if self.quantization_config.version == AWQLinearVersion.IPEX:
+        #     from ..integrations import post_init_awq_ipex_modules
+        #
+        #     model = post_init_awq_ipex_modules(model)
```
should this be removed ?

SunMarc reviewed (commented):
Thanks, left some minor comments !

Qubitium committed:
Merge branch 'main' into gptqmodel

ZX-ModelCloud committed:
CLEANUP

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
update test_awq

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
fix get_modules_to_not_convert()

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
fix test_awq.py::AwqTest::test_quantized_model_exllama

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

github-actions[bot] committed:
Apply style fixes

ZX-ModelCloud committed:
test_awq.py added EXPECTED_OUTPUT

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
update test_gptq.py

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

Qubitium committed:
Merge branch 'main' into gptqmodel

ZX-ModelCloud committed:
fix test_awq.py::AwqTest::test_save_pretrained

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

Qubitium commented on tests/quantization/autoawq/test_awq.py:
```diff
@@ -45,38 +41,38 @@
+    #     else:
+    #         # LLMAWQ does not work on a T4
+    #         with self.assertRaises(ValueError):
+    #             AwqConfig(bits=4, backend="llm-awq")
```
Fixed.

Qubitium commented:
@SunMarc Since last review:

1. Unused awq properties (fuse related) removed.
2. Fixed commented out code related to IPEX and change it to test TorchFused kernel instead (ipex replacement)
3. Fixed hf awq kernel selection not passing `device_map` to `hf_select_quant_linear_v2`. Without device_map, it was selecting the wrong kernel since gpt-qmodel needs device info to return best kernel for hw. 

The PR currently depends on GPT-QModel 5.4.4 which is not yet released as we are working to resolve asap an internal regression related to gptq packing code: https://github.com/ModelCloud/GPTQModel/issues/2234

ZX-ModelCloud committed:
use assertEqual() instead of assertTrue()

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
fix test_quantized_layers_class()

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
remove ExllamaV1 Test

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
format

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
fix get_modules_to_not_convert()

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
added EXPECTED_OUTPUT

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
remove ExllamaV1 Test

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
add AwqBackend.AUTO_TRAINABLE

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
Merge branch 'main' into gptqmodel

SunMarc committed:
Merge branch 'main' into gptqmodel

Qubitium committed:
Merge branch 'main' into gptqmodel

Qubitium commented:
> LMK when we can merge !

We are performing final CI run in gpt-qmodel and once that passes. I will push 5.6.0 release asap to pypi so this PR can be ready for merge.

Qubitium committed:
Merge branch 'main' into gptqmodel

SunMarc committed:
Merge branch 'main' into gptqmodel

SunMarc committed:
Merge branch 'main' into gptqmodel

MekkCyber commented on src/transformers/models/auto/tokenization_auto.py:
```diff
@@ -389,6 +389,8 @@ def load_merges(merges_file):
 
 def tokenizer_class_from_name(class_name: str) -> Union[type[Any], None]:
+    if class_name == "BloomTokenizer":
+        return TokenizersBackend
```
Not sure about this, the tokenizer should be fixed in a separate pr, and team members are already working on an other fix

ZX-ModelCloud replied:
Yes, I see other PRs are fixing this issue. I've already reverted it.

This is a temporary fix; otherwise, I can't load the Bloom Tokenizer.

SunMarc committed:
Merge branch 'main' into gptqmodel

MekkCyber reviewed (commented):
Thank you all for your work! I only have 2 nits

ZX-ModelCloud committed:
Update docs/source/zh/llm_tutorial.md

Co-authored-by: Mohamed Mekkouri <93391238+MekkCyber@users.noreply.github.com>

ZX-ModelCloud committed:
revert temporarily fix

Signed-off-by: ZX-ModelCloud <zx@modelcloud.ai>

ZX-ModelCloud committed:
Merge branch 'main' into gptqmodel

MekkCyber committed:
Merge branch 'main' into gptqmodel

MekkCyber committed:
Merge branch 'main' into gptqmodel

MekkCyber reviewed (approved):
Thank you for iterating lgtm

SunMarc committed:
Merge branch 'main' into gptqmodel

SunMarc committed:
Merge branch 'main' into gptqmodel

SunMarc reviewed (approved):
Thanks ! About gptq-model installation, would it be possible at some point to fetch the correct wheels depending on the user setup like torch using pip ? This is actually a quite a barrier for new users, or try to give more instructions ?

BenjaminBossan commented:
Just to be sure, this will be part of transformers v5?

Qubitium commented:
> 
> Okay, I must have a mismatched values hence it wasn't downloading the pre-compiled wheels. Is there a way to return a warning if this is not the case ? 

We missed building torch 2.9.1 whl for 5.6.0 so that is likely the cause. Will push 5.6.2 to resolve this other misc setup issues other users have reported plus logwarning when setup cannt match downloaded whl.
